source/rsa.c

```diff
@@ -0,0 +1,273 @@
+        return aws_raise_error(AWS_ERROR_CAL_BUFFER_TOO_LARGE_FOR_ALGORITHM);
+    }
+
+    return key_pair->vtable->encrypt(key_pair, algorithm, plaintext, out);
+}
+
+AWS_CAL_API int aws_rsa_key_pair_decrypt(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (AWS_UNLIKELY(ciphertext.len != (key_pair->key_size_in_bits / 8))) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Unexpected buffer size. For RSA, ciphertext is expected to match block size.");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    return key_pair->vtable->decrypt(key_pair, algorithm, ciphertext, out);
+}
+
+int aws_rsa_key_pair_sign_message(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    return key_pair->vtable->sign(key_pair, algorithm, digest, out);
+}
+
+int aws_rsa_key_pair_verify_signature(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    AWS_PRECONDITION(key_pair);
+
+    return key_pair->vtable->verify(key_pair, algorithm, digest, signature);
+}
+
+size_t aws_rsa_key_pair_block_length(const struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    return key_pair->key_size_in_bits / 8;
+}
+
+size_t aws_rsa_key_pair_signature_length(const struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    return key_pair->key_size_in_bits / 8;
+}
+
+int aws_rsa_key_pair_get_public_key(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_key_export_format format,
+    struct aws_byte_buf *out) {
+    (void)format; /* ignore format for now, since only pkcs1 is supported. */
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (key_pair->pub.len == 0) {
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    aws_byte_buf_init_copy(out, key_pair->allocator, &key_pair->pub);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_rsa_key_pair_get_private_key(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_key_export_format format,
+    struct aws_byte_buf *out) {
+    (void)format; /* ignore format for now, since only pkcs1 is supported. */
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (key_pair->priv.len == 0) {
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    aws_byte_buf_init_copy(out, key_pair->allocator, &key_pair->priv);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_der_decoder_load_private_rsa_pkcs1(struct aws_der_decoder *decoder, struct aws_rsa_private_key_pkcs1 *out) {
```

Comment:
I'm not sure how your decoder works, but at least on Linux/AWS-LC platforms you can use the built in one with [parse public key](https://github.com/aws/aws-lc/blob/main/include/openssl/rsa.h#L546-L548) and [parse private key](https://github.com/aws/aws-lc/blob/main/include/openssl/rsa.h#L567-L570)

Reply:
this is only used on windows to parse out components from the der and build bcrypt blob from it. Ill probably move this function to win impl file to avoid any confusion.

on mac we rely on Security framework functions to parse key from der. On linux we support building against openssl, aws-lc and boringssl. The 2 functions you linked are lc/boringssl specific, so i ended up relying on d2i_RSAPrivateKey, which provides similar functionality.